### PR TITLE
Search: Fix slow query when user does not have roles assigned

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What is this feature?**

This PR fixes slow dashboard search query in case of user with no roles assigned (None role).

**Why do we need this feature?**

When there's no roles assigned to the user, then search becomes very slow. There's an investigation in corresponding issue, but in general, we've got a shortcut for situation when there's no permissions, but it's broken after nested folders feature release.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-enterprise/issues/6523

### CODE REVIEW

**Review by papagian — APPROVED** [review]
LGTM

File: pkg/services/sqlstore/permissions/dashboard.go [hunk 1]
```diff
@@ -114,8 +114,21 @@ func (f *accessControlDashboardPermissionFilter) Where() (string, []any) {
 	return f.where.string, f.where.params
 }
 
+func (f *accessControlDashboardPermissionFilter) isEmptyPermissions() bool {
+	permissions := f.user.GetPermissions()
+	if len(permissions) == 1 {
```
Comment (IevaVasiljeva) [inline pkg/services/sqlstore/permissions/dashboard.go:119]: question: would the issue not occur with service accounts that have some permissions, but none of them are folder/dashboard permissions (eg, service account with permissions on data sources only).

File: pkg/services/sqlstore/permissions/dashboard.go [hunk 2]
```diff
@@ -114,8 +114,21 @@ func (f *accessControlDashboardPermissionFilter) Where() (string, []any) {
 	return f.where.string, f.where.params
 }
 
+func (f *accessControlDashboardPermissionFilter) isEmptyPermissions() bool {
```
Comment (IevaVasiljeva) [inline pkg/services/sqlstore/permissions/dashboard.go:117]: nit: it would be nice to add a comment describing why this "hack" is needed.

**Review by IevaVasiljeva — COMMENTED** [review]
LGTM, with a question to make sure that I understand in which cases the issue occurs.

File: pkg/services/sqlstore/permissions/dashboard.go (hunk 1 above)
Comment (alexanderzobnin) [reply pkg/services/sqlstore/permissions/dashboard.go:119]: Actually, yes, that's a good catch. I've found that the issue occurs when there are no dashboard/folder permissions, so maybe we could extend a check.

File: pkg/services/sqlstore/permissions/dashboard.go [hunk 3]
```diff
@@ -114,8 +114,21 @@ func (f *accessControlDashboardPermissionFilter) Where() (string, []any) {
 	return f.where.string, f.where.params
 }
 
+// Check if user has no permissions required for search to skip expensive query
+func (f *accessControlDashboardPermissionFilter) hasRequiredActions() bool {
```
Comment (alexanderzobnin) [inline pkg/services/sqlstore/permissions/dashboard.go:118]: @IevaVasiljeva This is a more accurate way to check all required permissions. We could skip "shared with me" in this case, it's folder permission, so it's only required when `/api/search?type=dash-folder` search is using, but I checked out, it's not affected by slow query bug.

**Review by IevaVasiljeva — APPROVED** [review]
Nice, this makes sense to me!

**Review by gamab — APPROVED** [review]

**Pull request merged**